core/src/main/java/io/grpc/internal/ServerCallImpl.java

```diff
@@ -221,6 +222,17 @@ public ServerStreamListenerImpl(
       this.call = checkNotNull(call, "call");
       this.listener = checkNotNull(listener, "listener must not be null");
       this.context = checkNotNull(context, "context");
+      // Wire ourselves up so that if the context is cancelled, our flag call.cancelled also
+      // reflects the new state. Use a DirectExecutor so that it happens in the same thread
+      // as the caller of {@link Context#cancel}.
+      this.context.addListener(
```

Comment:
I see that ClientCallImpl is slightly different, and instead implements cancelled(Context) by calling stream.cancel on the ClientStream. A few unit tests break if I do the analogous thing to the ServerStream. Is it worthwhile to follow the ClientCallImpl example and fix those unit tests? It didn't seem obvious to me, because it feels strange to receive a cancellation signal from the stream, surface it up to the level of a ServerCall, and then have the ServerCallImpl try to cancel the stream again. In the client case on the other hand, it seems sensible to initiate the cancellation on a ClientCall directly, and expect it to propagate it down to the ClientStream.